### PR TITLE
Bugfix: Parallel download of files

### PR DESCRIPTION
**Pull request by SSE4**
Changelog: Bugfix: Parallel download of files
Docs: omit
@TAGS: slow, svn
@PYVERS: Macos@py27, Windows@py36, Linux@py27, py34
related: https://github.com/conan-io/conan/issues/5477

benchmark is here: https://github.com/SSE4/conan-speed-test
parallelised function: `_download_files_to_folder` 

basically, it's getting about 2 times faster for me. `tqdm` progress bars work for me for parallel downloads (on OSX).

sequential (overral):
```
conan create . bincrafters/testing -k  8.30s user 3.77s system 5% cpu 4:00.04 total
conan create . bincrafters/testing -k  8.26s user 3.74s system 4% cpu 4:11.55 total
```
parallel (overral):
```
conan create . bincrafters/testing -k  9.13s user 4.71s system 10% cpu 2:12.07 total
conan create . bincrafters/testing -k  9.34s user 4.96s system 11% cpu 2:00.64 total
```

sequential (detailed):
```
boost_random/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_random/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_random/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 97.5kB/s]
Downloading conanfile.py: 100%|##########| 496/496 [00:00<00:00, 382kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 571kB/s]
AAA download files took: 0.755559921265
Decompressing conan_export.tgz: 766B [00:00, 195kB/s]
boost_random/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_base/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_base/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_base/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 80.8kB/s]
Downloading conanfile.py: 100%|##########| 10.1k/10.1k [00:00<00:00, 9.81MB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 589kB/s]
AAA download files took: 0.62668299675
Decompressing conan_export.tgz: 766B [00:00, 201kB/s]
boost_base/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_cycle_group_c/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_cycle_group_c/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_cycle_group_c/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 84.8kB/s]
Downloading conanfile.py: 100%|##########| 3.36k/3.36k [00:00<00:00, 3.29MB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 650kB/s]
AAA download files took: 1.91298604012
Decompressing conan_export.tgz: 766B [00:00, 674kB/s]
boost_cycle_group_c/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_algorithm/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_algorithm/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_algorithm/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 82.6kB/s]
Downloading conanfile.py: 100%|##########| 468/468 [00:00<00:00, 373kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 758kB/s]
AAA download files took: 0.639534950256
Decompressing conan_export.tgz: 766B [00:00, 707kB/s]
boost_algorithm/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_cycle_group_a/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_cycle_group_a/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_cycle_group_a/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 91.6kB/s]
Downloading conanfile.py: 100%|##########| 1.13k/1.13k [00:00<00:00, 931kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 582kB/s]
AAA download files took: 0.628607988358
Decompressing conan_export.tgz: 766B [00:00, 765kB/s]
boost_cycle_group_a/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_array/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_array/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_array/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 88.4kB/s]
Downloading conanfile.py: 100%|##########| 525/525 [00:00<00:00, 141kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 608kB/s]
AAA download files took: 0.628017187119
Decompressing conan_export.tgz: 766B [00:00, 758kB/s]
```
parallel (detailed):
```
boost_random/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_random/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_random/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 81.7kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 677kB/s]
Downloading conanfile.py: 100%|##########| 496/496 [00:00<00:00, 398kB/s]
AAA download files took: 0.747133016586
Decompressing conan_export.tgz: 766B [00:00, 200kB/s]
boost_random/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_base/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_base/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_base/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 132kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 857kB/s]
Downloading conanfile.py: 100%|##########| 10.1k/10.1k [00:00<00:00, 8.31MB/s]
AAA download files took: 0.325643062592
Decompressing conan_export.tgz: 766B [00:00, 669kB/s]
boost_base/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_cycle_group_c/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_cycle_group_c/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_cycle_group_c/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 199kB/s]
Downloading conanfile.py: 100%|##########| 3.36k/3.36k [00:00<00:00, 643kB/s]s]
AAA download files took: 0.212547063828
Decompressing conan_export.tgz: 766B [00:00, 608kB/s]
boost_cycle_group_c/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_algorithm/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_algorithm/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_algorithm/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 1.10MB/s]
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 119kB/s]
Downloading conanfile.py: 100%|##########| 468/468 [00:00<00:00, 361kB/s]
AAA download files took: 0.218121051788
Decompressing conan_export.tgz: 766B [00:00, 569kB/s]
boost_algorithm/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_cycle_group_a/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_cycle_group_a/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_cycle_group_a/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanfile.py: 100%|##########| 1.13k/1.13k [00:00<00:00, 1.00MB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 209kB/s]
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 79.6kB/s]
AAA download files took: 0.215751171112
Decompressing conan_export.tgz: 766B [00:00, 613kB/s]
boost_cycle_group_a/1.69.0@bincrafters/stable: Downloaded recipe revision 0
boost_array/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_array/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_array/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanfile.py: 100%|##########| 525/525 [00:00<00:00, 415kB/s]
Downloading conan_export.tgz: 100%|##########| 758/758 [00:00<00:00, 594kB/s]
Downloading conanmanifest.txt: 100%|##########| 103/103 [00:00<00:00, 79.6kB/s]
AAA download files took: 0.217633008957
Decompressing conan_export.tgz: 766B [00:00, 654kB/s]
```

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by uilianries** [comment]
From my side, I can say SSE4:parallel_download_files is faster.

Downloaded packages
```
Packages
    OpenSSL/1.0.2r@conan/stable:0abbb2ea17cdc92f4a2ac8a9e55de717e3b5a9d1 - Cache
    boost_algorithm/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_array/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_asio/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_assert/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_atomic/1.69.0@bincrafters/stable:46ef4796ad2bb84061a1e17f76c955feb1ab6174 - Download
    boost_bind/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_chrono/1.69.0@bincrafters/stable:c49547685d12323a82aa857582e8982b18726427 - Download
    boost_concept_check/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_config/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_container/1.69.0@bincrafters/stable:1660bcff8f85e20b285133f60de0c7ad7f3b0aee - Download
    boost_container_hash/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_context/1.69.0@bincrafters/stable:382b59306c616b65091eef123e54ab78974bbe8e - Download
    boost_conversion/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_core/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_coroutine/1.69.0@bincrafters/stable:6a29d9da6e42b3f64aed2c00fa58e534ea485449 - Download
    boost_cycle_group_a/1.69.0@bincrafters/stable:e158cb3aced07abc2eef4ec3fdc188f7e6e1748e - Download
    boost_cycle_group_b/1.69.0@bincrafters/stable:ab96835edcb44b4b736d0ed114c1477dbd1a6358 - Download
    boost_cycle_group_c/1.69.0@bincrafters/stable:1b51cf394c3df4fe42af75f9ae2d3be3d32d2c21 - Download
    boost_date_time/1.69.0@bincrafters/stable:76763034b12508e274f8e18c843f3429ab210cae - Download
    boost_detail/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_endian/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_exception/1.69.0@bincrafters/stable:cdd8f18f2e4c9eae29863bf73ece9c225a16daa1 - Download
    boost_filesystem/1.69.0@bincrafters/stable:12c9a5fa414002370de6af6050cdedb0ed59b5f3 - Download
    boost_foreach/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_function/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_function_types/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_fusion/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_integer/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_intrusive/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_io/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_iterator/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_lambda/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_lexical_cast/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_locale/1.69.0@bincrafters/stable:c77f7f6d920688ca9cb13ec27f5b496a7eec24c6 - Download
    boost_math/1.69.0@bincrafters/stable:1a3231f41ec5acf7f590cda11b9c5675e8804393 - Download
    boost_move/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_mpl/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_numeric_conversion/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_optional/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_phoenix/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_pool/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_predef/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_preprocessor/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_proto/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_random/1.69.0@bincrafters/stable:76763034b12508e274f8e18c843f3429ab210cae - Download
    boost_range/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_ratio/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_rational/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_regex/1.69.0@bincrafters/stable:037ada2ab43313bdb911fe596244065b51931852 - Download
    boost_smart_ptr/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_static_assert/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_system/1.69.0@bincrafters/stable:923f102c384aecf545487b4f4109a6d3fa5f9961 - Download
    boost_thread/1.69.0@bincrafters/stable:cb054d0b3e1ca595dc66bc2339d40f1f8f04ab31 - Download
    boost_throw_exception/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_tokenizer/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_tti/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_tuple/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_type_index/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_type_traits/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_typeof/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_unordered/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_utility/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_variant/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    boost_winapi/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    bzip2/1.0.6@conan/stable:76f87539fc90ff313e0b3182641a9bb558a717d2 - Cache
    cmake_findboost_modular/1.69.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    cpprestsdk/2.10.13@bincrafters/testing:6cbb2a9eefb5e849b2f0ba0647088ed895e3d110 - Build
    lzma/5.2.4@bincrafters/stable:76f87539fc90ff313e0b3182641a9bb558a717d2 - Cache
    websocketpp/0.7.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    zlib/1.2.11@conan/stable:76f87539fc90ff313e0b3182641a9bb558a717d2 - Cache
    zstd/1.3.5@bincrafters/stable:76f87539fc90ff313e0b3182641a9bb558a717d2 - Cache
```

Conan 1.17.0
```
real	5m48,745s
user	0m15,761s
sys	0m1,961s
```

@SSE4's branch
```
real	3m49,855s
user	0m15,898s
sys	0m2,256s
```

Full logs:

[conan-1.17.0.log](https://github.com/conan-io/conan/files/3403063/conan-1.17.0.log)
[SSE4-parallel_download_files.log](https://github.com/conan-io/conan/files/3403064/SSE4-parallel_download_files.log)

File: conans/client/rest/rest_client_v1.py [hunk 1]
```diff
@@ -29,6 +30,8 @@ def complete_url(base_url, url):
 
 class RestV1Methods(RestCommonMethods):
 
+    threadpool = ThreadPool(8)
```
Comment (uilianries) [inline conans/client/rest/rest_client_v1.py]: Why not cpu count?

Comment (SSE4) [reply conans/client/rest/rest_client_v1.py]: I think it serves very different purpose to configure parallel builds, and it's actually a number of processes. here we configure thread pool, which is within the same process.

**Comment by SSE4** [comment]
I believe timeout exceptions in tests is because I wait on the main thread, so it's blocked, and as mock server runs on the main thread as well, it's unable to process requests. I'll try to fix that.

**Comment by memsharded** [comment]
Interesting. I think it is worth to add the "@pyvers" specifier in the comment, and make sure this works for all python versions in all platforms without problems.

**Comment by SSE4** [comment]
@jgsogo I'll need some help on this one from the python expert. it turns out it might be a limitation of default `wsgiref` server implementation, which is a synchronous single-threaded server, and probably couldn't satisfy parallel requests (but it needs to be confirmed).

**Comment by jgsogo** [comment]
I think the problem could be related with your last comment, yes, but I'd need to have a closer look at it.

Nevertheless I'm bothered first with the `output`, the class `FileDownloader` that is used to download the files has some member variables, the one that concerns me now is the `output`, what happens if several threads write something, does the output get mixed?

**Comment by memsharded** [comment]
Outdated, only applied to V1, which has been removed in 2.0, so this no longer applies.

**Pull request closed without merging**